### PR TITLE
[Data] Account internal inqueue to previous operator's memory usage

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

To make informed scheduling decisions, Ray Data estimates the object store memory usage from each operator. Currently, an operator's inputs count towards its memory usage; however, this can lead to suboptimal scheduling decisions since we primarily care about outputs for backpressuring.

This PR updates the implementation so that input counts toward the previous operator's object store usage (i.e., in counts towards the operator that created the data).

## Related issue number

Required for https://github.com/ray-project/ray/pull/43026

See also https://github.com/ray-project/ray/pull/42851#discussion_r1472191429
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/execution/resource_manager.py [hunk 1]
```diff
@@ -49,13 +49,9 @@ def update_usages(self):
...
-            op_usage = op.current_resource_usage()
-            # Don't count input refs towards dynamic memory usage, as they have been
-            # pre-created already outside this execution.
-            if not isinstance(op, InputDataBuffer):
-                op_usage.object_store_memory = (
-                    op_usage.object_store_memory or 0
-                ) + state.outqueue_memory_usage()
+            op_usage = op.current_processor_usage()
+            assert not op_usage.object_store_memory
+            op_usage.object_store_memory = _estimate_object_store_memory(op, state)
```
Comment (raulchen) [inline python/ray/data/_internal/execution/resource_manager.py:54]: let's also add a unit test in `test_resource_manager.py` for this change

File: python/ray/data/tests/test_resource_manager.py [hunk 2]
```diff
@@ -138,3 +141,55 @@ def test_calculating_usage(self):
         assert resource_manager.get_downstream_object_store_memory(o1) == 1700
         assert resource_manager.get_downstream_object_store_memory(o2) == 1700
         assert resource_manager.get_downstream_object_store_memory(o3) == 1000
+
+    def test_object_store_usage(self):
```
Comment (raulchen) [inline python/ray/data/tests/test_resource_manager.py:145]: should add `restore_data_context` for this test.

File: python/ray/data/tests/test_executor_resource_management.py [hunk 3]
```diff
@@ -300,7 +360,8 @@ def test_actor_pool_resource_reporting_with_bundling(ray_start_10_cpus_shared):
         op.get_next()
 
     # Work is done, pool has been scaled down, and outputs have been consumed.
-    usage = op.current_resource_usage()
+    resource_manager.update_usages()
+    usage = resource_manager.get_op_usage(op)
```
[the remark this answers is not in the document]
Comment (raulchen) [reply python/ray/data/tests/test_executor_resource_management.py:364]: BTW, I plan to maintain a list of ops with accurate memory accounting. Then we'll only enable the memory reservation feature if the dataset only has these ops.

File: python/ray/data/tests/test_resource_manager.py (hunk 2 above)
Comment (bveeramani) [reply python/ray/data/tests/test_resource_manager.py:145]: Ah, good catch. 

Wonderiing why we don't just enable this by default for all tests?

**Review by raulchen — APPROVED** [review]

**Pull request merged**